internal/fs/fs.go

```diff
@@ -2154,10 +2154,16 @@ func (fs *fileSystem) OpenDir(
 
 	// Enables kernel list-cache in case of non-zero kernelListCacheTTL.
 	if fs.kernelListCacheTTL > 0 {
+		// Following lock ordering rules: first taking dirInode.RLock() and then fs.lock().
+		fs.mu.Unlock()
 
 		// Taking RLock() since ShouldInvalidateKernelListCache only reads the DirInode
 		// properties, no modification.
 		in.RLock()
```

Comment:
couple of questions:
1. prevDirListingTimestamp is getting updated in RmDir flow also. Is that correct?
2. Assuming that same directory list call will not be made if the data is already cached, do you need this locking here?
3. What happens if you read inconsistent data. Just trying to avoid the locking.

Reply:
1. Yes.
2. Once data is already in cache, then we only get OpenDir() and ReleaseDirHandle() call. No ReadDir() call. 
Although It makes sense to add the lock in OpenDir() but it doesn't help in reducing the no. of GCS List call. Since, OpenDir() will get completed before other's ReadDir(). So, taking a lock doesn't help here.
3. I don't see any harm in reading the inconsistent data. The same reasoning, OpenDir() gets completed before others ReadDir() call, so we would not be able to save the GCS List call by reading the consistent data. I am align with removing the lock from here. Making the change.

CC: @sethiay please let me know, if you want to discuss.

Reply:
One case to check is: What happens to the performance (duration of all lists to complete) if there are parallel lists on same directory with large number of files (say 1M flat directory) with and without this read lock ? IIUC, without this lock listing from GCS will happen twice and increase the duration ?

Reply:
Confirmed offline: lock will not save the GCS list call in this case because the readdir calls come for 2K entries at once. So, say there is a dir handle doing 1st readdir call and there is second call waiting on opendir read's lock. As soon as the first dir handle's first readdir's call complete, the opendir call will proceed and see that the prev update time is updated, however the kernel would not have cached the listing yet. After that the dir handle's second readdir call will come until the last last readdir and the opendir call of second reader will proceed to readdir making GCS calls.